File: exercises/exercises-petsc-p/matvec.py
#!/usr/bin/env python

import math
import sys

from petsc4py import init

init(sys.argv)
from petsc4py import PETSc  # noqa: E402

comm = PETSc.COMM_WORLD

nprocs = comm.Get_size()
procno = comm.Get_rank()


##
## Function to create the matrix
## Let's pretend that you do not know what happens here,
## so you have no information about the matrix size and such
##
def CreateMatrix(comm, nlocal):  # noqa: N802
    A = PETSc.Mat().create(comm=comm)  # noqa: N806
    A.setType(PETSc.Mat.Type.MPIAIJ)
    procno = comm.Get_rank()
    if procno == 0:
        nlocal = 1
    A.setSizes(((nlocal, None), (nlocal, None)))
    A.setPreallocationNNZ((3, 1))
    A.assemblyBegin(PETSc.Mat.AssemblyType.FLUSH)
    A.assemblyEnd(PETSc.Mat.AssemblyType.FLUSH)

    _, nglobal = A.getSize()
    first, last = A.getOwnershipRange()
    for row in range(first, last):
        col = row
        v = 2.0
        A.setValue(row, col, v)
        col = row - 1
        v = -1.0
        if col >= 0:
            A[row, col] = v
        col = row + 1
        v = -1
        if col < nglobal:
            A[row, col] = v
    A.assemblyBegin(PETSc.Mat.AssemblyType.FINAL)
    A.assemblyEnd(PETSc.Mat.AssemblyType.FINAL)
    return A


##
## Get a commandline argument for the size of the problem
##
nlocal = PETSc.Options().getInt("nlocal", 20)  # default is not optional

##
## Call the matrix creation routine
##
A = CreateMatrix(comm, nlocal)

A.view()  # just checking: screen output

#### Exercise:
#### - create a vector by finding the local size of the matrix
####
x = PETSc.Vec().create(comm=comm)
x.setType(PETSc.Vec.Type.MPI)
#### your code here ####
n_local_rows, _ = A.getLocalSize()
x.setSizes((n_local_rows, PETSc.DECIDE))
x.setFromOptions()
x.assemblyBegin()
x.assemblyEnd()
print("Vec local size on %d: %s" % (procno, str(x.getSizes())))

y = x.duplicate()
x.set(1)

####
#### Exercise:
#### - fill in the correct parameters to compute
####    the matrix vector product y = Ax
####
A.mult(x, y)  # your code here ####
# y.view()

nrm = y.norm(PETSc.NormType.NORM_1)
if abs(2 - nrm) > 1.0e-14:
    PETSc.Sys.Print("Wrong norm %e: should be 2." % nrm)
else:
    PETSc.Sys.Print("Global norm succeeds")

####
#### Second test, using local arrays
####
# localvec = PETSc.Vec().create(comm=PETSc.COMM_SELF)
# localvec.setType(PETSc.Vec.Type.SEQ)
# localsize = y.getLocalSize()
# localvec.setSizes(localsize)
#
# ####
# #### Exercise:
# #### -- put y's data into the local vector
# ####    and compute the local norm
# #### Hint:
# #### -- use GetArray to get the local data
# #### -- use PlaceArray to set it in the localvec
# ####
# #### your code here ####
#
# ####
# #### Compute the norm:
# #### should be 1 on first & last process,
# #### 0 elsewhere
# ####
# nrm = localvec.norm(PETSc.NormType.NORM_1)
# if procno in set(0, nprocs - 1):
#     if abs(nrm - 1.0) > 1.0e-14:
#         print("Local norm incorrect on %d: %e" % (procno, nrm))
#     else:
#         print("Local norm succeeds on %d" % procno)
# else:  # noqa: PLR5501
#     if abs(nrm) > 1.0e-14:
#         print("Local norm incorrect on %d: %e" % (procno, nrm))
#     else:
#         print("Local norm succeeds on %d" % procno)

File: exercises/exercises-petsc-c/matvec.c
/****************************************************************
 ****
 **** This program file is part of the tutorial
 **** `Introduction to the PETSc library'
 **** by Victor Eijkhout eijkhout@tacc.utexas.edu
 ****
 **** copyright Victor Eijkhout 2012-2020
 ****
 ****************************************************************/

static char help[] = "\nMatrix exercise. Commandline option: \"-nlocal 123\" for local block size.\n\n";

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#include <petscmat.h>
/*
 * Function to create the matrix.
 * Let's pretend that you do not know what happens here,
 * so you have no information about the matrix size and such.
 */
PetscErrorCode CreateMatrix(MPI_Comm comm, int nlocal, Mat* rA)
{

    Mat A;
    PetscInt first, last;
    PetscErrorCode ierr;
    int procno, nglobal;

    ierr = MatCreate(comm, &A);
    CHKERRQ(ierr);
    ierr = MatSetType(A, MATMPIAIJ);
    CHKERRQ(ierr);

    MPI_Comm_rank(comm, &procno);
    if (procno == 0)
        nlocal = 1;
    ierr = MatSetSizes(A, nlocal, nlocal, PETSC_DECIDE, PETSC_DECIDE);
    CHKERRQ(ierr);
    /*
     * Exercise
     * - what happens if you preallocate insufficient space?
     */
    ierr = MatMPIAIJSetPreallocation(A, 3, NULL, 1, NULL);
    CHKERRQ(ierr);

    ierr = MatAssemblyBegin(A, MAT_FLUSH_ASSEMBLY);
    CHKERRQ(ierr);
    ierr = MatAssemblyEnd(A, MAT_FLUSH_ASSEMBLY);
    CHKERRQ(ierr);

    ierr = MatGetSize(A, &nglobal, PETSC_NULLPTR);
    CHKERRQ(ierr);
    ierr = MatGetOwnershipRange(A, &first, &last);
    CHKERRQ(ierr);
    for (int row = first; row < last; row++)
    {
        PetscInt col  = row;
        PetscScalar v = 2.;
        ierr          = MatSetValue(A, row, col, v, INSERT_VALUES);
        CHKERRQ(ierr);
        col = row - 1;
        v   = -1.;
        if (col >= 0)
        {
            ierr = MatSetValue(A, row, col, v, INSERT_VALUES);
            CHKERRQ(ierr);
        }
        col = row + 1;
        v   = -1.;
        if (col < nglobal)
        {
            ierr = MatSetValue(A, row, col, v, INSERT_VALUES);
            CHKERRQ(ierr);
        }
    }
    ierr = MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY);
    CHKERRQ(ierr);
    ierr = MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY);
    CHKERRQ(ierr);

    *rA = A;
    return 0;
}

#undef __FUNCT__
#define __FUNCT__ "main"
int main(int argc, char** argv)
{
    PetscErrorCode ierr;
    ierr = PetscInitialize(&argc, &argv, PETSC_NULLPTR, help);
    CHKERRQ(ierr);

    MPI_Comm comm = PETSC_COMM_WORLD;
    int procno, nprocs;
    MPI_Comm_rank(comm, &procno);
    MPI_Comm_size(comm, &nprocs);

    /*
     * Get a commandline argument for the local block size of the problem
     */
    PetscInt nlocal = 20; // default local block size

    ierr = PetscOptionsGetInt(NULL, NULL, "-nlocal", &nlocal, NULL);
    CHKERRQ(ierr);

    /*
     * Call the matrix creation routine.
     * (Note the way we are treating the pointer to the matrix object.)
     */
    Mat A;
    ierr = CreateMatrix(comm, nlocal, &A);
    CHKERRQ(ierr);

    // ierr = MatView(A, PETSC_VIEWER_STDOUT_WORLD); // just checking : screen output
    // CHKERRQ(ierr);

    /*
     * Start creating a vector
     */
    Vec x;
    ierr = VecCreate(comm, &x);
    CHKERRQ(ierr);
    ierr = VecSetType(x, VECMPI);
    CHKERRQ(ierr);

    /*
     * Exercise:
     * -- find the local size of the matrix, and use that to set the vector size
     */
    {
        PetscInt rowsize;
        /**** your code here ****/
        ierr = MatGetLocalSize(A, &rowsize, PETSC_NULLPTR);
        CHKERRQ(ierr);
        /**** your code here ****/
        // ierr = VecSetSizes(x, rowsize, PETSC_DECIDE);
        ierr = VecSetSizes(x, rowsize, PETSC_DETERMINE);
        CHKERRQ(ierr);

        ierr = VecSetFromOptions(x);
        CHKERRQ(ierr);
    }

    /*
     * Duplicate some work vectors (of the same format and
     * partitioning as the initial vector).
     */
    Vec y;
    ierr = VecDuplicate(x, &y);
    CHKERRQ(ierr);

    /*
     * Set x to constant one
     */
    PetscScalar one = 1.0;

    ierr = VecSet(x, one);
    CHKERRQ(ierr);

    /*
     * Exercise:
     * -- fill in the correct parameters to compute
     *    the matrix vector product y = Ax
     */
    /**** your code here ****/
    ierr = MatMult(A, x, y);
    CHKERRQ(ierr);

    /*
     * First check on the product
     */
    double norm;
    ierr = VecNorm(y, NORM_1, &norm);
    CHKERRQ(ierr);

    if (fabs(2. - norm) > 1.e-14)
    {
        ierr = PetscPrintf(comm, "Wrong norm: %e should be 2.\n", norm);
        CHKERRQ(ierr);
    }
    else
    {
        ierr = PetscPrintf(comm, "Global norm test succeeds on all %d processes\n", nprocs);
        CHKERRQ(ierr);
    }

    // #if 0
    /*
     * Second, more funky test
     */
    // {
    //     PetscScalar* localdata;
    //     Vec localvec;
    //     int localsize;
    //     double norm, norm_shouldbe;
    //     /*
    //      * Create a single-process vector with the size of the local part of y
    //      */
    //     ierr = VecCreate(PETSC_COMM_SELF, &localvec);
    //     CHKERRQ(ierr);
    //     ierr = VecSetType(localvec, VECSEQ);
    //     CHKERRQ(ierr);
    //     ierr = VecGetLocalSize(y, &localsize);
    //     CHKERRQ(ierr);
    //     ierr = VecSetSizes(localvec, localsize, PETSC_DECIDE);
    //     CHKERRQ(ierr);
    //     /*
    //      * Exercise:
    //      * -- put y's data into the local vector,
    //      *    and compute the local norm
    //      * -- hint: use VecGetArray to get the y data
    //      * --       use VecPlaceArray to set that in localvec
    //      */
    //     /**** your code here ****/
    //     /*
    //      * Compute the norm: it should be 1 on the first & last process,
    //      * zero elsewhere
    //      */
    //     ierr = VecNorm(localvec, NORM_1, &norm);
    //     CHKERRQ(ierr);
    //     if (procno == 0 || procno == nprocs - 1)
    //         norm_shouldbe = 1.;
    //     else
    //         norm_shouldbe = 0.;
    //     if (fabs(norm - norm_shouldbe) > 1.e-14)
    //         printf("Wrong local norm %e on proc %d\n", norm, procno);
    //     else
    //         printf("Local norm succeeded\n");
    //
    //     /*
    //      * Let's clean up
    //      */
    //     /*
    //      * Optional exercise:
    //      * -- what are the mirror routines of VecGetArray & VecPlaceArray above?
    //      *    (this mostly serves to prevent memory leaks)
    //      */
    //     /**** your code here ****/
    //     ierr = VecDestroy(&localvec);
    //     CHKERRQ(ierr);
    // }
    // #endif

    /*
       Free work space.  All PETSc objects should be destroyed when they
       are no longer needed.
    */
    ierr = MatDestroy(&A);
    CHKERRQ(ierr);

    ierr = VecDestroy(&x);
    CHKERRQ(ierr);

    ierr = VecDestroy(&y);
    CHKERRQ(ierr);

    return PetscFinalize();
}
